### PR TITLE
incorporate into Q estimator Bill Premerlani's fast rotation handling from MatrixPilot

### PR DESCRIPTION
**Pull request by kd0aij**
small subset of https://github.com/PX4/Firmware/pull/4989

### CODE REVIEW

**Comment by LorenzMeier** [comment]
@kd0aij What is the status of this?

**Comment by kd0aij** [comment]
I'm waiting till https://github.com/PX4/Firmware/pull/5498 and https://github.com/PX4/jMAVSim/pull/36 are merged. Once those are in I'll be able to fly automated trajectories (attitude sequences) to thoroughly analyze performance, first in SITL with ground truth, then on a real vehicle.

**Comment by kd0aij** [comment]
rebased and tested in SITL and flight tested on QAV250. Significant reduction in attitude estimation errors during coordinated turns: (roll/pitch in degrees, note the good agreement with ground truth, especially after the turn stops)
fastrot_sitl_gt

This is now ready to merge.

**Comment by LorenzMeier** [comment]
Jenkins test this please.

**Comment by kd0aij** [comment]
for comparison, this is the same flight sequence without this commit: (not the significant pitch error after the turn stops)
banked_circle_lpe

**Comment by kd0aij** [comment]
flight tested and rebased on master

**Comment by jgoppert** [comment]
Looks good to merge to me.

**Comment by jgoppert** [comment]
Looks good to merge to me.

File: src/modules/attitude_estimator_q/attitude_estimator_q_main.cpp [hunk 1]
```diff
@@ -691,12 +699,13 @@ bool AttitudeEstimatorQ::update(float dt)
 	corr += (k % (_accel - _pos_acc).normalized()) * _w_accel;
 
 	// Gyro bias estimation
-	if (_gyro.length() < 1.0f) {
+	if (spinRate < 0.175f) {
```
Comment (jgoppert) [inline src/modules/attitude_estimator_q/attitude_estimator_q_main.cpp]: Do we know that this isn't the main reason for the performance improvement. It is a big differences on the magnitude of rotation rate at which we are estimating gyro biases from 1.0  (57 deg/2) to 0.175 (~6 deg/s).

**Comment by jgoppert** [comment]
I believe you, but I think there are two significant changes here and I don't swe any justification why we now only estimate gyro bias when we are rotating below 10 deg/s. I would be happy with just another sitl plot  vs ground truth to compare with those above when we set it back to 1 rad/s as the gyro bias estimation cut off.

**Comment by kd0aij** [comment]
Just out of curiosity, where did the 1 rad/sec value come from, originally? I don't remember seeing any testing that justified it.

**Comment by jgoppert** [comment]
@kd0aij I settled on adding the logic after doing some flight test and an a b test to see that it did help in high g turns. At the time, we didn't have any such logic so I wanted to be conservative. If we can afford to lower it to this value I'm fine with that.

**Comment by kd0aij** [comment]
OK, I'll repeat the testing I thought I had done earlier and document it properly this time. It's possible that the results could be different than with MatrixPilot anyway. One caveat I just remembered is that this needs to work with no GPS available for acceleration compensation.  I know I had the GPS disconnected during actual flight testing, but maybe not for SITL.

**Comment by kd0aij** [comment]
@jgoppert baseline Q estimator at yaw rate of about 0.5 rad/sec (significant bias error):
yawrate_0 5

and at about 0.2 rad/sec:
yawrate_0 2

**Comment by jgoppert** [comment]
Nice, I'm good to merge now.

**Comment by kd0aij** [comment]
@jgoppert I guess you have to do "another review approving the changes in the pull request" before this can be merged. That's actually pretty cool. I'll try not to make you mad at me :)

**Pull request merged**